docs/design/2023-09-22-tiflash-query-level-auto-spill.md

```diff
@@ -0,0 +1,55 @@
+
+* The memory usage threshold is per operator, so if a query contains multiple memory intensive operator, user have to set multiple threshold for each kind of operator
+* Even operator's memory usage threshold is set, there is no way to limit the overall memory usage at query level, and since there will be significant differences between queries, it is hard to set a suitable threshold for all kinds of queries
+
+In practise, a common need from user is that they wants to set a simple threshold to limit the overall memory usage for a query, and the database should auto trigger spill if the memory usage might exceeds the overall limit. It is hard or even not possible with current operator level spill framework. So we need to support automatic spill at query level.
+
+## Detailed Design
+
+In order to support query level auto spill, there are five basic problems to deal with
+
+### How to tracker memory usage at query level
+
+Currently, TiFlash already have a memory tracker framework, but, it tracks memory at the level of `MPPTask`, a query may have multiple `MPPTask`s, so in order to track memory usage at query level, we need to merge the memory tracker of all `MPPTask`s that belongs to the same query.
+A simplest way to do this is to let all `MPPTask`s use the same memory tracker. In order to do this, we need to put the memory tracker to a common place that can be seen by all the `MPPTask`s. In implementation,  we put the memory tracker in [`MPPQuery`](https://github.com/pingcap/tiflash/blob/ce24115c147293df10a9d9f1c745eb0e81f0f7b5/dbms/src/Flash/Mpp/MPPTaskManager.h#L75) . `MPPQuery` is a structure that contains all the `MPPTask`s for the same query, each `MPPTask` need to register to `MPPQuery`, and during register, it will get memory tracker from `MPPQuery`, so the memory tracker in `MPPQuery` can track all the memory usage for the `MPPQuery`
+
+### How to choose operators to spill
+
+In operator's spill, each operator will check the memory usage itself and trigger spill if the usage exceeds the threshold, but for query level auto spill, this trigger method does not work well because a query may have many operators that support spill, if each opeartor trigger spill by itself, it may ends up all the operator will spill data, which is not efficient and may make the query extremely slow. A Intuitive way to spill at query level is each time spill the operator that takes up most memories. In order to do this, there are serveral new data structures
+
+* [OperatorSpillContext](https://github.com/pingcap/tiflash/blob/ce24115c147293df10a9d9f1c745eb0e81f0f7b5/dbms/src/Core/OperatorSpillContext.h#L33), it is a structure that contains information about how many memories an operator may released when spill, for each operator that support spill, it will hold an `OperatorSpillContext`
```

Comment:
```suggestion
* [OperatorSpillContext](https://github.com/pingcap/tiflash/blob/ce24115c147293df10a9d9f1c745eb0e81f0f7b5/dbms/src/Core/OperatorSpillContext.h#L33), it is a structure that contains information about how many memories an operator may release when spill, for each operator that supports spill, it will hold an `OperatorSpillContext`
```